ebs/ebs.go

```diff
@@ -420,9 +421,107 @@ func (d *Driver) UpdateTags(volumeID string, newTags map[string]string) error {
 	return nil
 }
 
-func (d *Driver) FailoverLogic(volumeName string, volumeID string, opts map[string]string, newTags map[string]string, needsFS *bool) (string, int64, error) {
+func (d *Driver) BuildNewVolume(args *BuildArgs) (string, int64, error) {
+	volumeSize, err := d.getSize(args.opts, d.DefaultVolumeSize)
+	if err != nil {
+		return "", -1, err
+	}
+	volumeType, iops, err := d.getTypeAndIOPS(args.opts)
+	if err != nil {
+		return "", -1, err
+	}
+	r := &CreateEBSVolumeRequest{
+		Size:       volumeSize,
+		VolumeType: volumeType,
+		IOPS:       iops,
+		KmsKeyID:   d.DefaultKmsKeyID,
+		Encrypted:  d.DefaultEncrypted,
+	}
+	volumeID, err := d.ebsService.CreateVolume(r)
+	if err != nil {
+		return "", -1, err
+	}
+	log.Debugf("Created volume name=%v from EBS volume=%v", args.volumeName, volumeID)
+	return volumeID, volumeSize, nil
+}
+
+func (d *Driver) BuildVolumeFromScratch(volume *ec2.Volume, snapshot *ec2.Snapshot, args *BuildArgs) (string, int64, error) {
+	// If there is no current reference to the volumeName or the snapshot has opted out of failover the build a volume from scratch
+	if snapshot == nil && volume == nil {
+		log.Debugf("No recent snapshot or recent volume - Building name=%v from scratch", args.volumeName)
+		volumeId, volumeSize, err := d.BuildNewVolume(args)
+		return volumeId, volumeSize, err
+	} else if snapshot != nil && !shouldFailover(snapshot.Tags) {
+		// If there is a snapshot, but it opted out of failover then build from scratch
+		log.Debugf("A recent snapshot=%v exists for name=%v, but it has opted out of failover", *snapshot.SnapshotId, args.volumeName)
+		volumeId, volumeSize, err := d.BuildNewVolume(args)
+		return volumeId, volumeSize, err
+	} else if volume != nil && !d.isVolumeInLocalAz(volume) && !shouldFailover(volume.Tags) {
```

Comment:
you can make this third conditional a sub-conditional (nested) of the second one since it relies on the same for 2 out of 3 booleans.